src/storage/src/source/persist_source.rs

```diff
@@ -190,3 +194,218 @@ where
 
     (ok_stream, err_stream, token)
 }
+
+pub fn persist_source_listen_split<G>(
+    scope: &G,
+    persist_clients: Arc<Mutex<PersistClientCache>>,
+    metadata: CollectionMetadata,
+    as_of: Antichain<Timestamp>,
+) -> (
+    Stream<G, (Row, Timestamp, Diff)>,
+    Stream<G, (DataflowError, Timestamp, Diff)>,
+    Rc<dyn Any>,
+)
+where
+    G: Scope<Timestamp = mz_repr::Timestamp>,
+{
+    let worker_index = scope.index();
+    let peers = scope.peers();
+
+    let data_shard = metadata.data_shard.clone();
+
+    let persist_clients_moveable = persist_clients.clone();
+    let persist_location_moveable = metadata.persist_location.clone();
+    let as_of_moveable = as_of.clone();
+
+    // This source is split into two parts: a first part that sets up `async_stream` and a timely
+    // source operator that the continuously reads from that stream.
+    //
+    // It is split that way because there is currently no easy way of setting up an async source
+    // operator in materialize/timely.
+
+    // This is a generator that sets up an async `Stream` that can be continously polled to get the
+    // values that are `yield`-ed from it's body.
+    let async_stream = async_stream::try_stream!({
+        // Worker 0 is responsible for distributing splits from listen
+        if worker_index != 0 {
+            trace!("We are not worker 0, exiting...");
+            return;
+        }
+
+        let persist_client = persist_clients_moveable
+            .lock()
+            .await
+            .open(persist_location_moveable)
+            .await
+            .expect("could not open persist client")
+            .clone();
+
+        let mut read = persist_client
+            .open_reader::<SourceData, (), mz_repr::Timestamp, mz_repr::Diff>(data_shard.clone())
+            .await
+            .expect("could not open persist shard");
+
+        // Aggressively downgrade `since`, to not hold back compaction.
+        read.downgrade_since(as_of_moveable.clone()).await;
```

Comment:
I think this is copied over from an old impl of persist_source (this call got removed from persist_source at some point last week)

Reply:
Yeah, this hasn't been rebased yet bc I wanted to make sure the approach was sane before taking on that endeavor.